### ISSUE TITLE
[Bug]: Some `combine` works on Main thread

### ISSUE THREAD

**Issue by Jaehwa-Noh** [android/nowinandroid#1420]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is there a StackOverflow question about this issue?

- [X] I have searched StackOverflow

### What happened?

Some flow `combine` works on Main thread.

<br>

![image](https://github.com/android/nowinandroid/assets/48680511/2d49dc9f-6154-401a-89be-e76503149d9c)

![image](https://github.com/android/nowinandroid/assets/48680511/20b3bf5c-9006-4a22-b307-53be4efdad23)

![image](https://github.com/android/nowinandroid/assets/48680511/5e5f4de7-a9d7-476b-93d4-055c9d248eb4)

|![image](https://github.com/android/nowinandroid/assets/48680511/42812853-63b0-4459-ad5b-5ed5ca3291b3)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by yongsuk44** [comment]
Personally, umm... I'm not sure why `flowOn` needs to be added. I think that specifying a dispatcher in the coroutine builder would make thread management and tracking easier. 

If thread switching is implemented in the Repository and UseCase as you've modified, it seems that it could unnecessarily increase the amount of code to manage and make things more complicated.

**Comment by Jaehwa-Noh** [comment]
I added this `flowOn` on the Repository and UseCase because that `comebine` using almost all `map` method. `Map` would take long time when the collection become bigger, so that the logic need to be run on `Dispatchers.Default`.

**Comment by yongsuk44** [comment]
> I think that specifying a dispatcher in the coroutine builder would make thread management and tracking easier.

@Jaehwa-Noh 😀
This comment was left because I think that handling thread management and tracking with `flowOn` at the start of the coroutine builder or logic would be efficient.

```kotlin
// Starting point of logic
.flowOn(Dispatchers)
.stateIn(...)

// Coroutine builder
viewModelScope.launch(Dispatcher) { ... }
viewModelScope.async(Dispatcher) { ... }

```

Additionally, I also believe that performing complex computations on `Dispatchers.Default` is efficient. However, I'm not quite sure if it's efficient to switch the current tasks to `Default`. It would be effective if the project data size were very large, but in the current project, I'm unsure how much of a performance improvement there would be. Having a benchmark to compare before and after would make it easier to understand.  🤔

**Comment by Jaehwa-Noh** [comment]
@yongsuk44 I don't think that `repository` or `usecase` own coroutine dispatcher would be controlled in `viewModel` because `usecase` and `repository` could be re-used anywhere. Imagine that controlled by `coroutineScope` in the `viewModel`. When use that `repository` or `usecase` in another `viewModel`, developer must always consider that `viewModelScope` is working properly or not, and that leads them to be making mistake easily.

Current data size isn't enormous, but there's no reason we carry out the risk.
I don't say this is for improve the performance, I do this for prevent ANR, long calculate on main thread easily bring that error.

**Comment by yongsuk44** [comment]
When working with coroutines on `Dispatchers.Main`, there is no need to worry about ANR as long as the thread is not blocked. Most functions in our current project are designed as suspend functions, so they seem fine.

Of course, if there are incredibly complex calculations or tasks that block the thread, it would be reasonable to switch to `Dispatchers.Default` or `Dispatchers.IO`. However, these operations often involve calls to `Room` or `Retrofit`, most of which support the suspend modifier, so I think they should be fine.

However, as I mentioned in a previous comment, I'm not sure if adding `flowOn` and thereby increasing the amount of code we need to manage is worth the trade-off.

Am I missing anything here? 🤔

**Comment by Jaehwa-Noh** [comment]
@yongsuk44 You can check this [documentation](https://developer.android.com/kotlin/coroutines/coroutines-adv#main-safety).
>Dispatchers.Main - Use this dispatcher to run a coroutine on the main Android thread. This should be used only for interacting with the UI and performing quick work. Examples include calling suspend functions, running Android UI framework operations, and updating [LiveData](https://developer.android.com/topic/libraries/architecture/livedata) objects.

`Dispatchers.Main` is not safe for ANR. That is why `Room` and `Retrofit` using `Dispatchers.IO`.

**Comment by yongsuk44** [comment]
ANR occurs when the main thread is 'blocked' for a certain period of time. The suspension of coroutines and the blocking of threads are distinct concepts. 

Additionally, when implementing Room and Retrofit with the suspend modifier, the operations are handled on the IO.
